core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java

```diff
@@ -437,11 +437,43 @@ public Result visit(Filter e) {
       final Result x = visitInput(e, 0, Clause.WHERE);
       parseCorrelTable(e, x);
       final Builder builder = x.builder(e);
+      if (input instanceof Join) {
+        final Context context = x.qualifiedContext();
+        final ImmutableList.Builder<SqlNode> selectList = ImmutableList.builder();
+        // Fieldnames are unique since they are created by SqlValidatorUtil.deriveJoinRowType()
+        final List<String> uniqueFieldNames = input.getRowType().getFieldNames();
+        boolean selectListRequired = false;
+        for (int i = 0; i < context.fieldCount; i++) {
+          final SqlNode field = context.field(i);
+          final String fieldName = uniqueFieldNames.get(i);
+          if (fieldWasRenamedByJoinDueToNameClash(field, fieldName)) {
+            selectListRequired = true;
+          }
+          selectList.add(
+              SqlStdOperatorTable.AS.createCall(POS, field,
+              new SqlIdentifier(fieldName, POS)));
+        }
+        if (selectListRequired) {
+          builder.setSelect(new SqlNodeList(selectList.build(), POS));
+        }
+      }
       builder.setWhere(builder.context.toSql(null, e.getCondition()));
       return builder.result();
     }
   }
 
+  private static boolean fieldWasRenamedByJoinDueToNameClash(SqlNode field, String fieldName) {
+    if (!(field instanceof SqlIdentifier)) {
```

Comment:
context(AliasContext) field is only allowed to return SqlIdentifier.
and Is there any other simpler way to determine whether rename is needed?
can we identify the Duplicate name by Result.qualifiedContext().fieldList()

Reply:
I also thought about simplifying this on base of `Result.qualifiedContext().fieldList()`. But I had no idea how this could be achieved.

Reply:
Whether this method can satisfy？
```
private static boolean fieldRequiresAlias(Result x) {
  List<String> uniqueList = new ArrayList();
  for(SqlNode node : x.qualifiedContext().fieldList()) {
    if(node instance SqlIdentifier) { 
      SqlIdentifier field = (SqlIdentifier) node;
      if (uniqueList.contains(last(field.names))) {
          return true;
      }
      uniqueList.add(last(field.names));
    }
  }
  return false;
}
```

Reply:
I changed it accordingly.